core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java

```diff
@@ -1842,33 +1842,29 @@ public void delete(AlluxioURI path, DeleteContext context)
               LockPattern.WRITE_EDGE);
       try (LockedInodePath inodePath = mInodeTree
               .lockInodePath(lockingScheme)) {
-        mPermissionChecker.checkParentPermission(Mode.Bits.WRITE, inodePath);
+        mPermissionChecker.checkParentPermission(Mode.Bits.WRITE_EXECUTE, inodePath);
         mMountTable.checkUnderWritableMountPoint(path);
         if (!inodePath.fullPathExists()) {
           throw new FileDoesNotExistException(ExceptionMessage.PATH_DOES_NOT_EXIST
               .getMessage(path));
         }
 
+        List<String> failedChildren = new ArrayList<>();
         if (context.getOptions().getRecursive()) {
-          List<String> failedChildren = new ArrayList<>();
-          try (LockedInodePathList descendants = mInodeTree.getDescendants(inodePath)) {
-            for (LockedInodePath childPath : descendants) {
-              try {
-                mPermissionChecker.checkPermission(Mode.Bits.WRITE, childPath);
-                if (mMountTable.isMountPoint(childPath.getUri())) {
-                  mMountTable.checkUnderWritableMountPoint(childPath.getUri());
-                }
```

Comment:
The old implementation is very expensive at least in:
(1) It acquires all the locks
(2) The mount point check will generate a lot of temporary strings

Reply:
we acquire the locks in `deleteInternal()`. Were we iterating the entire descendant path twice?